### DISCUSSION TITLE
Skip List

### DISCUSSION THREAD

**Discussion by onee-only** [subak-ba/cs-study#41] [자료 구조]
# Skip List
스킵 리스트는 정렬된 원소들의 탐색과 삽입/삭제를 평균 `O(log n)` 의 시간복잡도를 가지는 **확률적** 자료 구조입니다. 

## 배열과 연결 리스트
아래는 정렬된 배열과 연결 리스트의 시간 복잡도의 비교입니다.

정렬된 배열:
- 탐색: `O(log n)` -> 이분탐색(이진탐색)
- 삽입/삭제: `O(n)` -> shifting 때문

연결 리스트:
- 탐색: `O(n)` -> 인덱스가 없기 때문에 head부터 순회
- 삽입/삭제: `O(1)` 

보시다시피 배열은 탐색은 빠르지만 수정은 느리고, 연결 리스트는 탐색은 느리지만 수정은 빠릅니다.
두 자료 구조 모두 각기 다른 장단점이 있기 때문에 trade-off가 발생합니다. 
스킵 리스트는 두 자료 구조의 장점만을 가져온 자료 구조입니다.

## 스킵 리스트의 구조
스킵 리스트를 이해하기 쉽도록 단계별로 스킵 리스트를 구축 해 보겠습니다.

### Level 1 (초기 상태)
스킵 리스트는 정렬된 연결 리스트에서 시작합니다. 이 층은 level 1으로 부릅니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/3e15698b-2b88-418f-af6b-2be806b941f6)

### Level 2
탐색 속도를 높히기 위해 각 원소의 다음 원소를 건너뛰는 새로운 지름길을 만듭니다. 
이렇게 만들어진 층은 level 2로 부릅니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/5691cc82-9f93-4d29-9009-63ebd86ff7cb)

### Level 3
탐색 속도를 더 높히기 위해 level 2의 각 원소들의 다음 원소를 건너뛰는 새로운 지름길을 만듭니다.
이렇게 만들어진 층은 level 3으로 부릅니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/6c7dd611-dd1a-4174-80ae-39e2756b022b)

## 스킵 리스트에서의 탐색
스킵 리스트에서의 탐색은 위에서 만든 지름길들을 활용하여 이루어집니다.
처음 커서는 가장 처음 원소의 가장 높은 레벨에서 시작합니다.
그리고 다음과 같은 작업을 반복합니다.

1. 만약 원소가 찾는 값이면 끝.
2. 만약 원소가 끝이 아니면서 원소의 현재 층의 포인터가 가리키고 있는 값이 더 작거나 같다면, 포인터가 가리키고 있는 원소로 이동하고 1번으로 이동.
3. 만약 그렇지 않다면, 커서의 층을 1층 내림.
4. 만약 커서의 층이 1층보다 작다면 끝. 아니면 1번으로 이동.

위 알고리즘을 바탕으로 위 리스트에서 8을 찾아 보면 다음과 같이 탐색합니다.

<img width="929" alt="image" src="https://github.com/subak-ba/cs-study/assets/103808587/5d5dc3a0-6c49-48f9-8ff4-96da1fa2a5ab">

위와 같이 8을 찾는 데 3번의 동작만 필요하게 되었습니다.
이러한 "완벽한" 형태의 스킵 리스트에서는 탐색이 최악의 상황에도 `O(log n)` 의 시간복잡도를 가지게 됩니다.

## 스킵 리스트의 무작위성
위와 같은 "완벽한" 형태의 스킵 리스트에서는 그 형태를 유지하기 위한 비용 때문에 삽입과 삭제의 성능이 떨어집니다.
그 대신, 무작위성을 추가하면 이 문제를 해결할 수 있습니다. 

<img width="471" alt="image" src="https://github.com/subak-ba/cs-study/assets/103808587/33531163-39ce-419f-8786-34abf9083f4e">

무작위성을 추가한 스킵 리스트는 위와 같이 생겼습니다. 더 이상 "완벽한" 형태를 유지하지 않습니다.

## 무작위성을 이용한 삽입과 삭제
스킵 리스트의 무작위성은 원소를 삽입할 때 활용됩니다. 
먼저, 원소를 삽입하기 위해서 원소가 들어갈 위치까지 탐색합니다.
원소가 삽입될 위치를 찾았으면 그 원소의 높이를 정해야 합니다. 

원소의 높이를 정하는 작업은 `동전 던지기`를 이용하여 정해집니다.
원소의 높이는 1에서 시작해서 동전을 던져 뒷면이 나올때까지 1 증가합니다.

그 후 그 원소의 전후 원소들의 포인터를 바꿔주면 됩니다. 탐색 과정에서 거쳐간 원소들을 저장 했다면 이 작업을 수월하게 할 수 있습니다.

![Skip List Add](https://github.com/subak-ba/cs-study/assets/103808587/8ee03988-7a0c-4c05-835c-ea4607eaf4b2)

스킵 리스트의 삭제는 삽입과 비슷하기 때문에 설명하지 않겠습니다.

## 스킵 리스트 복잡도 분석

각 동전 던지기에서 앞면이 나올 확률은 1/2이기 때문에, 원소가 i층(0층에서 시작한다고 가정)에 도달할 확률은 `( 1 / 2 ) ^ i`이 됩니다.
그렇게 되면, i층에 평균적으로 존재하는 원소의 개수는 원소가 n개일 때

> n * (( 1 / 2 ) ^ i ) = n / (2 ^ i)

위 공식을 다음과 같은 표가 나옵니다.

| 층 | 평균 원소 수 |
|--------|--------|
| 0 | n |
| 1 | n / 2 |
| 2 | n / 4 | 
| 3 | n / 8 | 
| i | n / (2 ^ i) | 

따라서 스킵 리스트의 공간을 모두 더하면

<img width="308" alt="image" src="https://github.com/subak-ba/cs-study/assets/103808587/1cee9107-daff-4158-9e54-b5ddcd5978be">

위에 따라 평균적인 공간복잡도는 `O(n)`이 됩니다. 최악의 경우에는 `O(n log n)` 이 될 수 있습니다.

스킵 리스트의 시간 복잡도를 알기 위해서는 수직과 수평의 탐색 거리를 알아야 합니다.

원소가 n개일 때 평균적으로 최대 높이는 `log n`이 나오게 됩니다. ( 최소한 원소의 개수가 1개인 층을 찾기 )
때문에 수직 탐색 시간 복잡도는 `O(log n)`이 됩니다.

i - 1층의 원소가 i층에도 존재할 확률은 1/2입니다. 
완벽한 스킵 리스트를 생각했을 때, 만약 1층에 원소가 8개 있다면 2층에는 4개가 있을 것입니다.
이 스킵 리스트에서는 각 층마다 최대 2개의 노드를 방문하면 원하는 값이 있는지 알 수 있을 것입니다.
만약 2개를 넘어가야 한다고 가정한다면, 그 노드는 이미 위 층에서 방문할 수 있었을 것입니다. 

![image](https://github.com/subak-ba/cs-study/assets/103808587/6c7dd611-dd1a-4174-80ae-39e2756b022b)

무작위성을 가진 스킵 리스트는 무조건 위 조건을 만족하지는 않지만, 확률로 봤을 때, 평균 2개의 노드를 거치면 수평 탐색이 끝나게 됩니다.
때문에 가장 윗 층에서부터 각 층마다 평균 2개의 노드만 방문하면 아래 층으로 내려갈 수 있는지 알 수 있게 됩니다.
그러므로 수평 탐색 시간 복잡도는 `O(1)`이 됩니다. 

따라서 스킵 리스트의 탐색은 최종적으로 `O(log n)`의 시간복잡도를 가지게 됩니다. 최악의 경우에는 O(n)이 될 수 있습니다.

## 이진 탐색 트리와 다른 점
이진 탐색 트리의 구현은 일반적으로 레드 블랙 트리가 이용됩니다.
레드 블랙 트리의 시간 복잡도도 `O(log n)` 이지만, 동시성을 추가하면 두 자료구조의 속도 차이가 생깁니다.

레드 블랙 트리에 삽입/삭제를 할 때는 트리 전체를 재조정 해야 하기 때문에 전체 트리에 락을 걸어야 합니다.
하지만 스킵 리스트에 삽입/삭제를 할 때에는 그 노드를 바라보고 있는 노드들에만 락을 걸면 됩니다.

스킵 리스트가 가지는 지역성 덕분에 스킵 리스트는 레드 블랙 트리보다 훨씬 빠르게 동작할 수 있다고 합니다.

스킵 리스트와 레드 블랙 트리의 동시성 관련하여 더 자세한 내용은 [스택 오버플로우 질문](https://stackoverflow.com/questions/256511/skip-list-vs-binary-search-tree)을 참고하시기 바랍니다.

## 레퍼런스
- [위키피디아](https://en.wikipedia.org/wiki/Skip_list)
- [유튜브 강의](https://www.youtube.com/watch?v=NDGpsfwAaqo)
- [brilliant](https://brilliant.org/wiki/skip-lists/)
- [티스토리](https://ohgym.tistory.com/10)

P.S: 스킵리스트 쫌 어렵네요.

**Comment by jangwooooo** [comment]
스킵리스트 재밌네요 ㅋㅋ  

<br>

> 그러므로 수평 탐색 시간 복잡도는 O(1)이 됩니다.  

근데 저 이 부분 이해가 잘 안돼요.
수평 탐색 시간의 경우 최악은 O(n)일 것 같은데 어떻게 O(1)를 가지는지 잘 모르겠어요

**Reply by onee-only** [reply]
보충 설명 해 놓겠다고 적어 놓고 안 해놨네요. 보충 해 놓겠습니다

**Reply by jangwooooo** [reply]
네 감사합니다

**Reply by onee-only** [reply]
> 수평 탐색 시간의 경우 최악은 O(n)일 것 같은데 어떻게 O(1)를 가지는지 잘 모르겠어요

스킵 리스트는 확률적 자료 구조이기 때문에 최악의 상황이라면 O(n)이 나올 수 있습니다.
하지만 그 확률이 극히 희박하기 때문에 평균 시간 복잡도인 O(1)을 사용합니다.

**Comment by jangwooooo** [comment]
동전던지기의 경우에는 어떤식으로 구현하나요?
`난수%2` 이런식으로 구현하나요?

**Reply by onee-only** [reply]
이론적인 스킵 리스트에서 동전 던지기(coin flip)의 확률(probability)은 p라는 변수를 통해 정해지는데, 이 수는 0 ~ 1 까지의 값을 가집니다.
이를 통해 랜덤한 0부터 1의 값과 p를 비교하여 성공인지 실패인지 정하면 됩니다. 

만약 `랜덤값 > p`를 성공으로 봤을 때, 보통 p는 1/2 혹은 1/4로 사용한다고 합니다.
이런 가정에서,
p가 0에 가까워질 수록 원소가 윗 층에 올라갈 확률이 줄어 탐색 시간이 늘고,
p가 1에 가까워질 수록 반대로 윗 층에 올라갈 확률이 늘어 메모리 공간을 더 사용하게 됩니다.

구현체 중에서는 부동소수점 연산의 속도를 고려하여 p의 값을 정수로 두기도 합니다.

**Reply by jangwooooo** [reply]
답변 감사합니다.
그러면 p의 값이 1/4일 경우 0에 가까우니 윗 층에 올라갈 확률이 줄고, p의 값이 4/3일 경우 1에 가까우니 윗 층에 올라갈 확률이 는다고 이해했습니다.
제가 이해한 게 맞나요?

**Reply by onee-only** [reply]
3/4 말하시는 거라면 넵 맞습니다

**Reply by jangwooooo** [reply]
아 오타났네요 ㅋㅋ..
답변감사합니다 👍
